utilities/src/signals.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2019-2025 Provable Inc.
+// This file is part of the snarkOS library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::{
+    Arc,
+    atomic::{AtomicBool, Ordering},
+};
+use tokio::sync::Notify;
+
+use tracing::{debug, error};
+
+/// Generic trait that can be queried for whether current process should be stopped.
+/// This is implemented by `SignalHandler` and `SimpleStoppable`.
+pub trait Stoppable: Send + Sync {
+    /// Initiates shutdown of the node.
+    fn stop(&self);
+
+    /// Returns `true` if the node is (in the process of being) stopped.
+    fn is_stopped(&self) -> bool;
+}
+
+/// Wrapper around `AtomicBool` that implements the `Stoppable` trait.
+///
+/// This is useful when no signal or complex shutdown handling is necessary (e.g., in a test environment).
+pub struct SimpleStoppable {
+    state: AtomicBool,
+}
+
+impl SimpleStoppable {
+    pub fn new() -> Arc<Self> {
+        Arc::new(Self { state: AtomicBool::new(false) })
+    }
+}
+
+impl Stoppable for SimpleStoppable {
+    fn stop(&self) {
+        self.state.store(true, Ordering::SeqCst);
+    }
+
+    fn is_stopped(&self) -> bool {
+        self.state.load(Ordering::SeqCst)
+    }
+}
+
+/// Helper for signal handling that implements the `Stoppable` trait.
+///
+/// This struct will set itself to "stopped" as soon as the process receives Ctrl+C.
+/// It can also be manually stopped (e.g., when the node encounters a fatal error).
+pub struct SignalHandler {
+    stopped: AtomicBool,
+    notify: Notify,
+}
+
+impl SignalHandler {
+    /// Spawns a background tasks that listens for Ctrl+C and returns `Self`.
+    pub fn new() -> Arc<Self> {
+        let obj = Arc::new(Self { stopped: AtomicBool::new(false), notify: Default::default() });
+
+        {
+            let obj = obj.clone();
+            tokio::spawn(async move {
+                obj.handle_signals().await;
+            });
+        }
```

Comment:
nit: I'd move the spawning to `handle_signals`, so as not to confuse object creation with doing additional work